`OnTheFlyMerging` fails for Horizontal Diffusion if `GreedyMerging` is turned off.

If  in the gtc backends, `GreedyMerging` is turned off, the tests fail in the `OnTheFlyMerging` optimization pass with
```
src/gtc/passes/oir_optimizations/horizontal_execution_merging.py:257: in visit_Stencil
    return oir.Stencil(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for Stencil
E   __root__
E     Symbols ['res', 'res', 'res'] not found. (type=value_error)

pydantic/main.py:362: ValidationError

```
If `OnTheFlyMerging` is turned off as well, the tests pass.

--- a/src/gtc/passes/oir_optimizations/horizontal_execution_merging.py
+++ b/src/gtc/passes/oir_optimizations/horizontal_execution_merging.py
@@ -124,7 +124,6 @@ class OnTheFlyMerging(NodeTranslator):
         self,
         horizontal_executions: List[oir.HorizontalExecution],
         symtable: Dict[str, Any],
-        tmps_to_remove: Set[str],
         new_symbol_name: Callable[[str], str],
     ) -> List[oir.HorizontalExecution]:
         """Recursively merge horizontal executions.
@@ -181,10 +180,9 @@ class OnTheFlyMerging(NodeTranslator):
             or first_has_large_body()
             or first_has_expensive_function_call()
         ):
-            return [first] + self._merge(others, symtable, tmps_to_remove, new_symbol_name)
+            return [first] + self._merge(others, symtable, new_symbol_name)
 
         writes = first_accesses.write_fields()
-        tmps_to_remove |= writes
         others_otf = []
         for horizontal_execution in others:
             read_offsets: Set[Tuple[int, int, int]] = set()
@@ -223,7 +221,7 @@ class OnTheFlyMerging(NodeTranslator):
                 )
             others_otf.append(merged)
 
-        return self._merge(others_otf, symtable, tmps_to_remove, new_symbol_name)
+        return self._merge(others_otf, symtable, new_symbol_name)
 
     def visit_VerticalLoopSection(
         self, node: oir.VerticalLoopSection, **kwargs: Any
@@ -233,30 +231,28 @@ class OnTheFlyMerging(NodeTranslator):
             horizontal_executions=self._merge(node.horizontal_executions, **kwargs),
         )
 
-    def visit_VerticalLoop(
-        self, node: oir.VerticalLoop, *, tmps_to_remove: Set[str], **kwargs: Any
-    ) -> oir.VerticalLoop:
+    def visit_VerticalLoop(self, node: oir.VerticalLoop, **kwargs: Any) -> oir.VerticalLoop:
         if node.loop_order != common.LoopOrder.PARALLEL:
             return node
-        sections = self.visit(node.sections, tmps_to_remove=tmps_to_remove, **kwargs)
+        sections = self.visit(node.sections, **kwargs)
+        accessed = AccessCollector.apply(sections).fields()
         return oir.VerticalLoop(
             loop_order=node.loop_order,
             sections=sections,
-            caches=[c for c in node.caches if c.name not in tmps_to_remove],
+            caches=[c for c in node.caches if c.name in accessed],
         )
 
     def visit_Stencil(self, node: oir.Stencil, **kwargs: Any) -> oir.Stencil:
-        tmps_to_remove: Set[str] = set()
         vertical_loops = self.visit(
             node.vertical_loops,
             symtable=node.symtable_,
             new_symbol_name=symbol_name_creator(set(node.symtable_)),
-            tmps_to_remove=tmps_to_remove,
             **kwargs,
         )
+        accessed = AccessCollector.apply(vertical_loops).fields()
         return oir.Stencil(
             name=node.name,
             params=node.params,
             vertical_loops=vertical_loops,
-            declarations=[d for d in node.declarations if d.name not in tmps_to_remove],
+            declarations=[d for d in node.declarations if d.name in accessed],
         )

